### ISSUE TITLE
Compiling with MSVC Community Edition 2017

### ISSUE THREAD

**Issue by at67** [lb3361/gigatron-lcc#2]
These are changes I had to make to get glcc to build:

lcc.c:
~~~
#ifndef _MSC_VER
extern int access(char *, int);
extern int getpid(void);
#endif
~~~

cpp.h and cpp.c:
 I had to rename all instances of outp to outp_, outp clashes with a the Windows CRT outp function.

CMakeLists.txt:
I replaces all instances of py -3 to python3, py -3 can be flakey in choosing python3 on some windows installations.
Added escaped quotes, (\"), to list(APPEND objlist "${obj}") and list(APPEND objlist "${fn}") so that they look like this:
list(APPEND objlist "\"${obj}\"")
list(APPEND objlist "\"${fn}\"")
I have spaces in some of my folder names.

I also had to manually copy these exes from build/Release to build: (CMake on Windows/MSVC by default stores exes into Debug/Release folders).
~~~
cpp.ese
gtsim.exe
lburg.exe
lcc.exe
rcc.exe
~~~

This is as far as I got, all executables build correctly, but the executing of lcc.exe for libc4, libc5, libc6, libconx and libsim fails with a:
~~~
error MSB6001: Invalid command line switch for "cmd.exe". Illegal characters in path.
~~~

**Comment by lb3361** [comment]
I've made a number of changes.

1. The suggested changes to the cpp source have been made.
2. The CMakeLists now invoke FindPython3() and use the detected python path instead of the launcher py. I am not sure that this is perfect, but one can always change 'Python3_EXECUTABLE' with ccmake.
3. The suggested escapes to handle white space in filenames broke the cmake Unix makefile generator. Instead I refactored the CMakeLists to only deal with relative filenames in foreach commands. The full path prefix is constructed as late as possible and never in a foreach. 
4. The MSB6001 error is mysterious. Old versions of cmake had a bug that caused this message. I am afraid that all my add_custom_{command|target} might be still buggy.

Let me know when you get a chance to try again...

**Comment by at67** [comment]
I forgot to mention yesterday that I had to add **_typedef long off_t;_** to gtsim.c, so that it looks like this, (to get it to compile for MSVC, MSVC doesn't have the off_t typedef and instead uses long as the return type for functions like lseek()):

This same change is needed, (or something similar), for this current repo.
*edit* You might want to add a _MSC_VER guard around it.
~~~
#ifndef WIN32
# include <unistd.h>
# include <fcntl.h>
#else
# include <io.h>
# include <fcntl.h>
# define read _read
# define write _write
# define lseek _lseek
# define close _close
# define fileno _fileno
# define dup2 _dup2
#ifdef _MSC_VER
typedef long off_t;
#endif
#endif
~~~

Also I still had to manually copy the exe files from build/Release to build, see above.

Now the build fails with this: (which is what I was trying to fix with the escaped quotes yesterday, I have a space between Gigatron and TTL in my path).
~~~
13>lcc.exe: can't find `TTL/glcc/build/CMakeFiles/libsim.glccdir/_iob_setup_4.o'
13>lcc.exe: can't find `D:/Projects/Gigatron'
13>lcc.exe: can't find `TTL/glcc/gigatron/mapsim/libsim/_iob_setup.c'
12>lcc.exe: can't find `TTL/glcc/build/CMakeFiles/libconx.glccdir/cons_geom_4.o'
12>lcc.exe: can't find `D:/Projects/Gigatron'
12>lcc.exe: can't find `TTL/glcc/gigatron/mapconx/libconx/cons_geom.c'
9>lcc.exe: can't find `TTL/glcc/build/CMakeFiles/libc4.glccdir/_avladd_4.o'
9>lcc.exe: can't find `D:/Projects/Gigatron'
9>lcc.exe: can't find `TTL/glcc/gigatron/libc/_avladd.c'
11>lcc.exe: can't find `TTL/glcc/build/CMakeFiles/libc6.glccdir/_avladd_6.o'
11>lcc.exe: can't find `D:/Projects/Gigatron'
11>lcc.exe: can't find `TTL/glcc/gigatron/libc/_avladd.c'
10>lcc.exe: can't find `TTL/glcc/build/CMakeFiles/libc5.glccdir/_avladd_5.o'
10>lcc.exe: can't find `D:/Projects/Gigatron'
10>lcc.exe: can't find `TTL/glcc/gigatron/libc/_avladd.c'
~~~
It looks like your list of objects, (objlist), is still being confused by the space in my path and are being passed through argv to lcc as semi-colon AND space separated arguments. I'm guessing the arguments to lcc.exe need to be contained in quotes for paths with potential spaces.

**Comment by at67** [comment]
Just tried the latest version:
1) exes are now in the correct folder.
2) gtsim.c still needs typedef long off_t; to compile, (see above).
3) New errors are shown below:
~~~
17>Generating CMakeFiles/libsim.glccdir/_iob_setup_4.o
16>Generating CMakeFiles/libconx.glccdir/cons_geom_4.o
14>Generating CMakeFiles/libc-cpu5.glccdir/_avladd_5.o
15>Generating CMakeFiles/libc-cpu6.glccdir/_avladd_6.o
13>Generating CMakeFiles/libc-cpu4.glccdir/_avladd_4.o
17>cpp: Can't open input file TTL\glcc\build/cpp
16>cpp: Can't open input file TTL\glcc\build/cpp
13>cpp: Can't open input file TTL\glcc\build/cpp
14>cpp: Can't open input file TTL\glcc\build/cpp
15>cpp: Can't open input file TTL\glcc\build/cpp
~~~
It looks like the paths were truncated at the space in "D:/Projects/Gigatron TTL/glcc/build/"

**Comment by lb3361** [comment]
Working on it.  I've setup a virtual machine with msvc2019...

- L.

⁣Get BlueMail for Android ​

**Comment by lb3361** [comment]
I believe the latest version (commit 92d80197) works with MSVC.

The problem was not space in filenames but a sort of bug in either the visual studio cmake generator or in msbuild.
In order to cross compile each runtime library (e.g. libc), the previous CMakeLists.txt had one add_custom_command() for each file and an add_custom_target() for the library itself.  This used to work very well with the makefile generators. But when using visual studio, msbuild would only execute one of those commands (the first one with an inexistent target file) instead of all the library dependencies. I tried several ways to organize the custom commands and targets without changing that. It may be that msbuild breaks when a target has too many custom file dependencies (I only tries MSVC 2019).

The solution was to completely change how we cross-compile. Instead of trying to force it into the build system, I generate a cmake script that calls the compiler and concatenates its outputs. The library target then only invokes this script. This is less smart about tracking the dependencies, but it seems to be far more reliable.

**Comment by at67** [comment]
I'm getting this now: (I'll have a proper look at it after work)
~~~
15>Cross-compiling CMakeFiles/libconx.glccdir/cons_geom_4.o
16>Cross-compiling CMakeFiles/libsim.glccdir/_iob_setup_4.o
14>Cross-compiling CMakeFiles/libc-cpu6.glccdir/_avladd_6.o
12>Cross-compiling CMakeFiles/libc-cpu4.glccdir/_avladd_4.o
13>Cross-compiling CMakeFiles/libc-cpu5.glccdir/_avladd_5.o
14>cpp: Can't open input file TTL\glcc\build/cpp
15>cpp: Can't open input file TTL\glcc\build/cpp
13>cpp: Can't open input file TTL\glcc\build/cpp
12>cpp: Can't open input file TTL\glcc\build/cpp
16>cpp: Can't open input file TTL\glcc\build/cpp
12>CMake Error at CMakeFiles/libc-cpu4.glccdir/libc-cpu4.cmake:16 (file):
12>  file failed to open for reading (No such file or directory):
12>
12>    D:/Projects/Gigatron TTL/glcc/build/CMakeFiles/libc-cpu4.glccdir/_avladd_4.o
12>
12>
12>C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(209,5): error MSB6006: "cmd.exe" exited with code 1.
12>Done building project "libc-cpu4.vcxproj" -- FAILED.
13>CMake Error at CMakeFiles/libc-cpu5.glccdir/libc-cpu5.cmake:16 (file):
13>  file failed to open for reading (No such file or directory):
13>
13>    D:/Projects/Gigatron TTL/glcc/build/CMakeFiles/libc-cpu5.glccdir/_avladd_5.o
13>
13>
13>C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(209,5): error MSB6006: "cmd.exe" exited with code 1.
13>Done building project "libc-cpu5.vcxproj" -- FAILED.
14>CMake Error at CMakeFiles/libc-cpu6.glccdir/libc-cpu6.cmake:16 (file):
14>  file failed to open for reading (No such file or directory):
14>
14>    D:/Projects/Gigatron TTL/glcc/build/CMakeFiles/libc-cpu6.glccdir/_avladd_6.o
14>
14>
14>C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(209,5): error MSB6006: "cmd.exe" exited with code 1.
14>Done building project "libc-cpu6.vcxproj" -- FAILED.
16>CMake Error at CMakeFiles/libsim.glccdir/libsim.cmake:16 (file):
16>  file failed to open for reading (No such file or directory):
16>
16>    D:/Projects/Gigatron TTL/glcc/build/CMakeFiles/libsim.glccdir/_iob_setup_4.o
16>
16>
16>C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(209,5): error MSB6006: "cmd.exe" exited with code 1.
16>Done building project "libsim.vcxproj" -- FAILED.
15>CMake Error at CMakeFiles/libconx.glccdir/libconx.cmake:16 (file):
15>  file failed to open for reading (No such file or directory):
15>
15>    D:/Projects/Gigatron TTL/glcc/build/CMakeFiles/libconx.glccdir/cons_geom_4.o
15>
15>
15>C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(209,5): error MSB6006: "cmd.exe" exited with code 1.
15>Done building project "libconx.vcxproj" -- FAILED.
17>------ Rebuild All started: Project: ALL_BUILD, Configuration: Release Win32 ------
17>Building Custom Rule D:/Projects/Gigatron TTL/glcc/CMakeLists.txt
18>------ Skipped Rebuild All: Project: INSTALL, Configuration: Release Win32 ------
18>Project not selected to build for this solution configuration 
========== Rebuild All: 12 succeeded, 5 failed, 1 skipped ==========
~~~

**Comment by lb3361** [comment]
```
14>cpp: Can't open input file TTL\glcc\build/cpp
```

This does not seem to be in cmake but in the compiler driver, lcc....

**Comment by lb3361** [comment]
What a mess.

GLCC uses the normal LCC compiler driver `lcc.c`.  Under windows, `lcc.c` uses function `_spawnvp()` to execute other processes such as `cpp`, `rcc`, or the linker.  It turns out that `_spawnwp()` simply collects its arguments separated by white spaces and passes that to `CreateProcess()`. So if one of the arguments contains a space, the receiving process will interpret it as two arguments. 

Therefore
```
char *cpp =  "D:/Projects/Gigatron TTL/glcc/build/cpp"
char *args[] = { cpp,  "somesource.c", "somefile.i", 0 };
 _spawnvp(P_WAIT, cpp, args );
```
executes `cpp` correctly. But `cpp` then interprets the arguments as 
```
"D:/Projects/Gigatron" "TTL/glcc/build/cpp" "somesource.c" "somefile.i"
```
and this is why it is complaining that file `"TTL/glcc/build/cpp"` does not exist. 

Conclusion: I had to write a wrapper for `_spawnvp()` in `lcc.c` that quotes the arguments as necessary. This is far trickier than expected because various windows programs have subtle differences in their argument parsing code when it comes to properly escaping a double quote inside a double quoted string. I had to research existing code bases for the safest way to do this. See commit https://github.com/lb3361/gigatron-lcc/commit/81f6cd02

Then I hit a bug in gnu-make which also forgets to quote arguments. I reported it to the gnu-make people at  http://savannah.gnu.org/bugs/index.php?61154 .  I think this bug does not affect the build process though.

Anyway, I would again like to know how this works for you.

**Comment by at67** [comment]
That builds perfectly now, just these 3 warnings :)
~~~
10>../src/gigatron.md(1910): warning C4113: 'void (__cdecl *)()' differs in parameter lists from 'void (__cdecl *)(int,int,int,int)'
10>../src/gigatron.md(1910): warning C4113: 'void (__cdecl *)()' differs in parameter lists from 'void (__cdecl *)(int,int,int,int,int,int *)'
10>../src/gigatron.md(1910): warning C4113: 'void (__cdecl *)()' differs in parameter lists from 'void (__cdecl *)(Node)'
~~~

**Comment by at67** [comment]
I ran your test program in gtsim.exe and everything works fine, but when I run the a.gt1 file in my emulator, (with your dev.rom), it crashes at the SYS $fd, as the SYS routine at $ff01 is not real code.
~~~
$0ead  LDI    $00        : 0xfe : 0xd2f1 : 0x46f0 : 0x3160 : 0x5b13 : 0x75ea : 0xeacf : 0x275e : 0xeacf : 0x265e : 0xeacf : 0x285e
$0eaf  BEQ    $da        : 0xfe : 0xd2f1 : 0x46f0 : 0x3160 : 0x5b13 : 0x75ea : 0xeacf : 0x275e : 0xeacf : 0x265e : 0xeacf : 0x285e
$0eda  LDWI   $ff01      : 0xfe : 0xd2f1 : 0x46f0 : 0x3160 : 0x5b13 : 0x75ea : 0xeacf : 0x275e : 0xeacf : 0x265e : 0xeacf : 0x285e
$0edd  STW    $22        : 0xfe : 0xd2f1 : 0x46f0 : 0x3160 : 0x5b13 : 0x75ea : 0xeacf : 0x275e : 0xeacf : 0x265e : 0xeacf : 0x285e
$0edf  SYS    $fd        : 0xfe : 0xd2f1 : 0x46f0 : 0x3160 : 0x5b13 : 0x75ea : 0xeacf : 0x275e : 0xeacf : 0x265e : 0xeacf : 0x285e
~~~

So what SYS routine is stored at $ff01 in the ROM that my emulator is not finding? When I disassemble the ROM at 0xff01 all I see is gibberish. Is gtsim patching the ROM before it executes the gt1 files?

P.S. My emulator provides a vCPU trace whenever a timing error occurs or non code is executed, that's how I grabbed this code.

P.P.S. That LDI 0/BEQ instruction pair seem superfluous :P

Ok I had a look at gtsim.c, pretty cool, but you'll need to document it somewhere prominently so that people don't make the same mistake I did and think that the .gt1 file produced will work in any emulator or on hardware.

You should probably use a different extension, (other than .gt1), for these types of gigatron executables as documented by Marcel in docs/GT1-files.txt, i.e.
~~~
"By publishing a file as GT1 you're signaling that you expect the file can be loaded with any past, present
or future ROM version. To help remind the programmer which features were introduced after ROM v1,
and therefore require a romType check before use, we'll add a ROM version number to their names."
~~~

**Comment by at67** [comment]
Let me know when you are happy with the state of glcc and I'll re-organise the gigatron root to deprecate the old lcc and add glcc as the defacto c compiler.

P.S. It looks great to me and is more than ready for inclusion IMHO.

**Comment by lb3361** [comment]
When you compile with -map=sim, you get a special version of the stdio functions that communicate with gtsim in order to make it read, write, or even open files on their behalf.  This is how text printed from the gt1 appears on the gtsim stdout instead of the non-existant gtsim screen.  This communication is achieved by means of pseudo sys calls in range ff00-ff10, directly captured by the emulator (no rom patch needed.)  In fact the emulator also captures SYS_Exec in order to run the gt1 specified on the command line instead of MainMenu and SYS_SetMode 

See https://github.com/lb3361/gigatron-lcc/blob/81f6cd02708792201f1847329fdee39a19dbd03d/gigatron/mapsim/gtsim.c#L176 and https://github.com/lb3361/gigatron-lcc/blob/81f6cd02708792201f1847329fdee39a19dbd03d/gigatron/mapsim/gtsim.c#L602


Basically, this means that a program that uses stdio and is compiled with -map=sim works only in gtsim. It neither works in a real gigatron nor in another emulator,

I should make that clearer in the doc.
Other than that, I am quite happy with the state of glcc.  I am sure bugs will show up, but nothing as final as with the old lcc.

- Leon




⁣Get BlueMail for Android ​

**Comment by lb3361** [comment]
Yes.

I just created a branch named GLCC_BRANCH_1.3 for that purpose.  Maybe this is a good case for 'git subtree'  ( https://www.atlassian.com/git/tutorials/git-subtree ) 

- Leon




⁣Get BlueMail for Android ​
